libs/sync/files_state.go

```diff
@@ -0,0 +1,103 @@
+package sync
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/databricks/cli/libs/fileset"
+)
+
+// FilesState keeps track of files on the local filesystem and their corresponding
+// entries in WSFS.
+type FilesState struct {
+	// Map of local file names to their last recorded modified time. Files found
+	// to have a newer mtime have their content synced to their remote version.
+	LastModifiedTimes map[string]time.Time `json:"last_modified_times"`
+
+	// Map of local file names to corresponding remote names.
+	// For example: A notebook named "foo.py" locally would be stored as "foo"
+	// in WSFS, and the entry would be: {"foo.py": "foo"}
+	LocalToRemoteNames map[string]string `json:"local_to_remote_names"`
+
+	// Inverse of LocalToRemoteNames. Together they form a 1:1 mapping where all
+	// the remote names and local names are unique.
+	RemoteToLocalNames map[string]string `json:"remote_to_local_names"`
+}
+
+// Convert an array of files on the local file system to a FilesState representation.
+func toFilesState(ctx context.Context, localFiles []fileset.File) (*FilesState, error) {
+	fs := &FilesState{
+		LastModifiedTimes:  make(map[string]time.Time),
+		LocalToRemoteNames: make(map[string]string),
+		RemoteToLocalNames: make(map[string]string),
+	}
+
+	// Expect no files to have a duplicate entry in the input array.
+	seen := make(map[string]fileset.File)
+	for _, f := range localFiles {
+		if _, ok := seen[f.Relative]; !ok {
+			seen[f.Relative] = f
+		} else {
+			return nil, fmt.Errorf("expected only one entry per file. Found duplicate entries for file: %s", f.Relative)
+		}
+	}
+
+	// Compute the new state.
+	for _, f := range localFiles {
+		fs.LastModifiedTimes[f.Relative] = f.Modified()
+		remoteName, err := f.WsfsPath()
+		if err != nil {
+			return nil, err
+		}
+
+		if existingLocalName, ok := fs.RemoteToLocalNames[remoteName]; ok {
+			return nil, fmt.Errorf("both %s and %s point to the same remote file location %s. Please remove one of them from your local project", existingLocalName, f.Relative, remoteName)
+		}
+
+		fs.LocalToRemoteNames[f.Relative] = remoteName
+		fs.RemoteToLocalNames[remoteName] = f.Relative
+	}
+	return fs, fs.validate()
+}
+
+// Consistency checks for the sync files state representation. These are invariants
+// that downstream code for computing changes to apply to WSFS depends on.
+//
+// Invariants:
+//  1. All entries in LastModifiedTimes have a corresponding entry in LocalToRemoteNames
+//     and vice versa.
+//  2. LocalToRemoteNames and RemoteToLocalNames together form a 1:1 mapping of
+//     local <-> remote file names.
+func (fs *FilesState) validate() error {
```

Comment:
Why do we need to do this? FilesState which we test in not read from state file but built inside `toFilesState` function. So in the end  It just seems like we're unit testing `toFilesState` code in this validate method.

Reply:
Yeah that's true, never thought about it that way!

One good reason to do this is that these invariants are critical to correct operation of the syncer. Violating these can cause very hard to debug issues.
Also this is better than a unit test because this validates all real world use cases.